### ISSUE TITLE
机器学习实战：从数据预处理到模型部署

### ISSUE THREAD

**Issue by sbumsid24** [sbumsid24/XH#4]
1. 数据预处理与特征工程
1.1 数据清洗与转换
import pandas as pd
import numpy as np
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# 加载数据集
data = pd.read_csv('data.csv')

# 查看数据基本信息
print(data.info())
print(data.describe())

# 处理缺失值
def handle_missing_values(df):
    # 数值型列用中位数填充
    num_cols = df.select_dtypes(include=['int64', 'float64']).columns
    num_imputer = SimpleImputer(strategy='median')
    df[num_cols] = num_imputer.fit_transform(df[num_cols])
    
    # 类别型列用众数填充
    cat_cols = df.select_dtypes(include=['object']).columns
    cat_imputer = SimpleImputer(strategy='most_frequent')
    df[cat_cols] = cat_imputer.fit_transform(df[cat_cols])
    
    return df

# 处理异常值
def handle_outliers(df, column):
    Q1 = df[column].quantile(0.25)
    Q3 = df[column].quantile(0.75)
    IQR = Q3 - Q1
    
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    df[column] = np.where(df[column] < lower_bound, lower_bound, df[column])
    df[column] = np.where(df[column] > upper_bound, upper_bound, df[column])
    
    return df

# 特征工程流水线
def create_feature_pipeline():
    # 数值特征处理
    numeric_features = ['age', 'income', 'credit_score']
    numeric_transformer = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='median')),
        ('scaler', StandardScaler())
    ])
    
    # 类别特征处理
    categorical_features = ['gender', 'education', 'marital_status']
    categorical_transformer = Pipeline(steps=[
        ('imputer', SimpleImputer(strategy='most_frequent')),
        ('onehot', OneHotEncoder(handle_unknown='ignore'))
    ])
    
    # 组合预处理步骤
    preprocessor = ColumnTransformer(
        transformers=[
            ('num', numeric_transformer, numeric_features),
            ('cat', categorical_transformer, categorical_features)
        ])
    
    return preprocessor

# 执行数据预处理
cleaned_data = handle_missing_values(data)
for col in ['age', 'income', 'credit_score']:
    cleaned_data = handle_outliers(cleaned_data, col)

preprocessor = create_feature_pipeline()
X = preprocessor.fit_transform(cleaned_data)
2. 机器学习模型构建与训练
2.1 分类模型实现
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report, accuracy_score
from sklearn.model_selection import GridSearchCV

# 准备数据
y = cleaned_data['target']
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42)

# 随机森林模型
rf_model = RandomForestClassifier(random_state=42)
rf_params = {
    'n_estimators': [100, 200, 300],
    'max_depth': [None, 10, 20],
    'min_samples_split': [2, 5, 10]
}

rf_grid = GridSearchCV(rf_model, rf_params, cv=5, n_jobs=-1)
rf_grid.fit(X_train, y_train)
best_rf = rf_grid.best_estimator_

# SVM模型
svm_model = SVC(probability=True, random_state=42)
svm_params = {
    'C': [0.1, 1, 10],
    'kernel': ['linear', 'rbf'],
    'gamma': ['scale', 'auto']
}

svm_grid = GridSearchCV(svm_model, svm_params, cv=5, n_jobs=-1)
svm_grid.fit(X_train, y_train)
best_svm = svm_grid.best_estimator_

# 逻辑回归模型
lr_model = LogisticRegression(random_state=42, max_iter=1000)
lr_params = {
    'C': [0.1, 1, 10],
    'penalty': ['l1', 'l2'],
    'solver': ['liblinear']
}

lr_grid = GridSearchCV(lr_model, lr_params, cv=5, n_jobs=-1)
lr_grid.fit(X_train, y_train)
best_lr = lr_grid.best_estimator_

# 模型评估
def evaluate_model(model, X_test, y_test):
    y_pred = model.predict(X_test)
    print(f"Model: {type(model).__name__}")
    print(f"Accuracy: {accuracy_score(y_test, y_pred):.4f}")
    print(classification_report(y_test, y_pred))
    print("="*60)

evaluate_model(best_rf, X_test, y_test)
evaluate_model(best_svm, X_test, y_test)
evaluate_model(best_lr, X_test, y_test)
2.2 模型集成与优化
from sklearn.ensemble import VotingClassifier, StackingClassifier
from sklearn.model_selection import cross_val_score

# 投票集成
voting_clf = VotingClassifier(
    estimators=[
        ('rf', best_rf),
        ('svm', best_svm),
        ('lr', best_lr)
    ],
    voting='soft'
)

voting_clf.fit(X_train, y_train)
evaluate_model(voting_clf, X_test, y_test)

# 堆叠集成
stacking_clf = StackingClassifier(
    estimators=[
        ('rf', best_rf),
        ('svm', best_svm)
    ],
    final_estimator=LogisticRegression(),
    cv=5
)

stacking_clf.fit(X_train, y_train)
evaluate_model(stacking_clf, X_test, y_test)

# 交叉验证评估
models = {
    'Random Forest': best_rf,
    'SVM': best_svm,
    'Logistic Regression': best_lr,
    'Voting Classifier': voting_clf,
    'Stacking Classifier': stacking_clf
}

for name, model in models.items():
    scores = cross_val_score(model, X, y, cv=5, scoring='accuracy')
    print(f"{name} CV Accuracy: {scores.mean():.4f} (±{scores.std():.4f})")
3. 模型部署与API服务
3.1 使用Flask部署模型
import joblib
from flask import Flask, request, jsonify
import numpy as np

# 保存最佳模型和预处理管道
joblib.dump(best_rf, 'model/rf_model.pkl')
joblib.dump(preprocessor, 'model/preprocessor.pkl')

# 创建Flask应用
app = Flask(__name__)

# 加载模型和预处理管道
model = joblib.load('model/rf_model.pkl')
preprocessor = joblib.load('model/preprocessor.pkl')

@app.route('/predict', methods=['POST'])
def predict():
    try:
        # 获取请求数据
        data = request.get_json()
        
        # 转换为DataFrame格式
        input_data = pd.DataFrame([data])
        
        # 数据预处理
        processed_data = preprocessor.transform(input_data)
        
        # 模型预测
        prediction = model.predict(processed_data)
        proba = model.predict_proba(processed_data)
        
        # 返回结果
        return jsonify({
            'prediction': int(prediction[0]),
            'probability': 
https://qqmww.com/baike/3412.html
https://qqmww.com/baike/3413.html
https://qqmww.com/baike/3414.html
https://qqmww.com/baike/3415.html
https://qimingziw.com/baike/2945.html
https://qimingziw.com/baike/2946.html
https://qimingziw.com/baike/2947.html
https://qimingziw.com/baike/2948.html
https://qimingziw.com/baike/2949.html
https://qingganhui.com/baike/3343.html
https://qingganhui.com/baike/3344.html
https://qingganhui.com/baike/3345.html
https://qingganhui.com/baike/3346.html
https://qingganhui.com/baike/3347.html
https://qqmww.com/xinwen/nba/3427.html
https://qqmww.com/xinwen/nba/3428.html
https://qqmww.com/xinwen/nba/3439.html
https://qqmww.com/xinwen/nba/3440.html
https://qqmww.com/xinwen/nba/3441.html
